### ISSUE TITLE
MD3 to Unreal

### ISSUE THREAD

**Issue by DarkinMan** [acqu/rtcw-wet-blender-model-tools#14]
Hi,

It's not really an issue, but rather an help request. 

I'd like to export the head's model, but when I import the result file into my game engine, the model is splitted into 4 parts ("Teeth", "Eye01", "Eye02", "Head"), and here's the issue. First, it miss the "Tag_Mouth" part, and second, in the game engine, for some reasons, animations aren't working anymore, nothing is moving, even if I manually link the animations to their assigned skeleton parts. 

I only have basic knowledges about 3D Modelling, so unfortunatly, I can't find any fix for that.

Thanks in advance for the help you could provide me.

**Comment by mino-git** [comment]
Hi, what happens if you export and then immediately import into blender again. Are animations and tags preserved? And which game engine are you using?

**Comment by DarkinMan** [comment]
So, the tags are here, but the animations aren't working. I can see the animation data on the timeline, but nothing's moving. And I'm using Unreal Engine 4.

**Comment by mino-git** [comment]
Again, please be precise, otherwise i can't help.

Answer this question: "Hi, what happens if you export and then immediately import into blender again."

**Comment by DarkinMan** [comment]
No, I mean in Blender. I did exactly what you said, and this is the result I have, into blender.

**Comment by mino-git** [comment]
Correct me if i am wrong. So you say the animations are not working in blender after you export from blender (to .md3), and then import that exported file into blender again. Cause that would mean something is wrong with the import export script.

As for unreal: it natively supports .obj and .fbx files. Make sure to export to .fbx from blender (not .obj), and animations and tags should be imported there too. EDIT this is because .obj does not support animation and tag objects.

**Comment by mino-git** [comment]
Also, you might join our discord if you want: https://discord.gg/uQ3cGm97 Lots of unreal users there.

**Comment by DarkinMan** [comment]
I'll explain you the process I followed :

1- Start a new scene in blender, removing base light, camera, and cube
2- Import via your plugin the Head.md3 file (with working animations at this point)
3- Apply the shading of the skinfile I want (I did with axis_soldier.skin)
4- Export the scene as "Axis Soldier Animated Head.fbx", with "Bake Animations" ticked (Same settings than what I did for the bodies, which are working perfectly).
5- Import this fbx into both UE4 or Blender
6- Animations are not working anymore, even if I can see that the animation data is still present.
7- (Only on UE4): "Head", "Teeth", "Eye01", and "Eye02" are now splitted into different files, with their own animation sequences (but not moving anything on the mesh they're attached to).

I joined the Discord server as Darkin Man#9707, if it's easier / fatser for you to talk with me

**Comment by mino-git** [comment]
Hey there, now i see and you are right, i get the same results.

For background info: https://docs.blender.org/manual/en/latest/animation/shape_keys/introduction.html

My importer for .md3 uses "absolute shape keys" for animation. After the import you should see all animations and tags correct. But when using the blender .fbx exporter (or other exporters like collada) the result is indeed no animations. I tried different options, but nothing worked. It seems the blender exporters simply do not read these kind of information.

Since i am interested in this too i leave this open. idTech 3 to Unreal workflow is kind of important to me too. Maybe if someone else got a solution i am all ear. Maybe if you convert the absolute shape keys to relative shape keys it could trigger blenders exporters to finally read out the animation data.

Again, thanks for the report and maybe a solution will pop up sometime :)

**Comment by mino-git** [comment]
Also if you don't mind, i renamed the issue to "MD3 to Unreal".

**Comment by DarkinMan** [comment]
I'm thinking about it since a few hours now, and I was wondering. Is there any way to simply split the .md3 to seperate mdm/mdx files and then import them the same way as the body, which seems to work fine ?

**Comment by gzmuszynski** [comment]
I see I wasn't the only one with the idea. did it went anywhere?

**Comment by DarkinMan** [comment]
Nothing came up since that, sorry :/

**Comment by mino-git** [comment]
Yeah, unfortunately no solution.

A good starting point may be animation info the .fbx exporters read.

As far as i remember the discussion on discord, we tried figuring out how to convert shape key information (animation info) into something the standard exporters of blenders read and export to.

**Comment by mino-git** [comment]
As a sidenote: the skeletal based formats work well. You import into Blender, export to .fbx, done (with minor config).

For md3 or mdc, which use vertex based animation, this is not the case. And the reason is, as stated already, but i think i did not explain it that well from my perspective as an addon writer: the addon writes the animation data as single key frames into 'special' data fields of blender. These data fields are simply not read by the .fbx exporters of blender. However, it must certainly be possible to write the animation data to other 'special' data fields, from where they are read from the exporters of blender. And that is the point: i simply know no other way of writing this data during import process. If i would know where to write it in the blender data fields other than what is done currently (which works 'perfectly' well for the et rtcw use cases), then i could rewrite the addon so an import export bridge from md3 / mdc to fbx through blender is possible.

It may even be possible with other toolchains currently, i just have not seen it yet. Yeah long story short: i would not know where to write the data during import for this specific use case (unreal use case). Would appreciate any hints.

**Comment by Thunderdojo** [comment]
Have you tried to make a skeleton with a root bone and then bones NOT attached but still has the root as main and then do some constraint to the respective tag, and then export as a skeletal fbx file?
that might work I can actually try in the weekend and see

**Comment by mino-git** [comment]
Unfortunately i have not UE installed, so i can not test. But worth a try.

**Comment by mino-git** [comment]
Found some useful links:

- https://blender.stackexchange.com/questions/291237/cant-bake-vertex-animation-to-object-when-export-as-fbx
- https://dev.epicgames.com/documentation/en-us/unreal-engine/importing-animations-using-fbx-in-unreal-engine
- https://blenderartists.org/t/scripting-converting-vertex-animation-to-skeletal-animation-via-shape-keys/1364095

I am bit confused, does fbx support vertex animation or not? And then: does Unreal even support vertex based anims?

**Comment by Thunderdojo** [comment]
no it doesnt support that, I will give it a test later today

**Comment by mino-git** [comment]
So basically an export bridge from blender to .fbx would be possible, but since UE does not support vertex based anims in .fbx the import to Unreal via .fbx will fail.

What would also work is to import to blender, and recreate the animations frame by frame, with a custom armature, like suggested in one of the links. There is even a request for a plugin to generate that data automatically. Anyone knows such an addon?

Then when you have the skeletal mesh + anims, import will go just as usual in UE.
